### PR TITLE
Get id token claims

### PR DESCRIPTION
**Pull request by 7opf**
By submitting a PR to this repository, you agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md). Please see the [contributing guidelines](https://github.com/auth0/.github/blob/master/CONTRIBUTING.md) for how to create and submit a high-quality PR for this repo.

### Description

Added idTokenClaims$ observable to AuthServiece to expose the underlying getIdTokenClaims() function of auth0-spa.js in the same way as user$ observable.

Copied the user$ implementation and tests, replaced relevant bits with idTokenClaim.

Regenerated the docs with npm run build

### Testing

> Describe how this can be tested by reviewers. Be specific about anything not tested and reasons why. If this library has unit and/or integration testing, tests should be added for new functionality and existing tests should complete without errors.
>
> Please include any manual steps for testing end-to-end or functionality not covered by unit/integration tests.
>
> Also include details of the environment this PR was developed in (language/platform/browser version).

- [x] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [x] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used, if not `master`

### CODE REVIEW

**Review by lbalmaceda — COMMENTED** [review]
@7opf Please, revert the last commit as the documentation is generated as part of the release process and here is just adding noise to the diff. Someone from the team will pick up later. Thanks!

**Comment by 7opf** [comment]
> @7opf Please, revert the last commit as the documentation is generated as part of the release process and here is just adding noise to the diff. Someone from the team will pick up later. Thanks!

OK, I've reverted. Thank you.

**Comment by stevehobbsdev** [comment]
Thanks for this @7opf - I've rebased to master and added a commit that allows our integration tests to run with a dummy user.

Let me give it a quick test and get back to you 👍

File: projects/auth0-angular/src/lib/auth.service.spec.ts [hunk 1]
```diff
@@ -136,6 +137,34 @@ describe('AuthService', () => {
...
+      (auth0Client.getIdTokenClaims as jasmine.Spy).and.resolveTo(claims);
+
+      service.idTokenClaims$.subscribe((value) => {
+        expect(value).toBe(claims);
+        done();
+      });
+    });
+
+    it('should get the ID token claims if not authenticated', (done) => {
+      (auth0Client.isAuthenticated as jasmine.Spy).and.resolveTo(true);
```
Comment (frederikprijck) [inline projects/auth0-angular/src/lib/auth.service.spec.ts]: I know the existing tests are identical (like for getUser), but why does the test mention `if not authenticated` but does use `resolveTo(true)`?

File: projects/playground/e2e/integration/playground.spec.ts [hunk 2]
```diff
@@ -51,14 +45,13 @@ describe('Smoke tests', () => {
   it('do redirect login and show user and access token', () => {
     cy.visit('/');
     cy.get('#login').should('be.visible').click();
-
     cy.url().should('include', 'https://brucke.auth0.com/login');
     loginToAuth0();
-
     cy.get('[data-cy=userProfile]').contains(`"email": "${EMAIL}"`);
-
+    cy.get('[data-cy=idTokenClaims]').contains('__raw');
```
Comment (frederikprijck) [inline projects/playground/e2e/integration/playground.spec.ts:51]: Would it be worth checking something else? Like email ?

**Review by frederikprijck — CHANGES_REQUESTED** [review]

File: projects/playground/e2e/integration/playground.spec.ts (hunk 2 above)
Comment (stevehobbsdev) [reply projects/playground/e2e/integration/playground.spec.ts:51]: We can do if you feel it would add value. I just check `__raw` as it only ever appears in the ID token claims. I just wanted a sanity check to make sure the value was being returned.

File: projects/auth0-angular/src/lib/auth.service.spec.ts (hunk 1 above)
Comment (stevehobbsdev) [reply projects/auth0-angular/src/lib/auth.service.spec.ts]: Good spot, something doesn't seem right here. If a user wasn't authenticated I would not expect to be able to get the claims at all. That observable should not emit￼ anything.

Comment (7opf) [reply projects/auth0-angular/src/lib/auth.service.spec.ts]: I was wondering this also. I think the test should read `should not get the ID token claims if not authenticated` and `resolveTo(false)`.

Comment (frederikprijck) [reply projects/auth0-angular/src/lib/auth.service.spec.ts]: It might also be trying to test that if the user is authenticated, but there are no claims, it still emits something being falsey.  In that case, the test is correct but the description is wrong. 

I am not sure that is a super useful test tho. So maybe reworking it as @7opf mentions is a good idea.

File: projects/playground/e2e/integration/playground.spec.ts (hunk 2 above)
Comment (frederikprijck) [reply projects/playground/e2e/integration/playground.spec.ts:51]: I am fine with what we have. Was just curious if checking some value being present adds more value instead of checking the existence of a property. But let's keep it as is.

File: projects/auth0-angular/src/lib/auth.service.spec.ts (hunk 1 above)
Comment (7opf) [reply projects/auth0-angular/src/lib/auth.service.spec.ts]: I've updated the negative tests for user$ and idTokenClaims$ in latest commit

File: projects/auth0-angular/src/lib/auth.service.spec.ts [hunk 3]
```diff
@@ -126,14 +127,65 @@ describe('AuthService', () => {
...
+      service.user$.subscribe((v) => {
+        value = v;
+      });
+
+      setTimeout(() => {
+        expect(value).toBeUndefined();
+        done();
+      }, 1000);
+    });
+  });
```
Comment (stevehobbsdev) [inline projects/auth0-angular/src/lib/auth.service.spec.ts]: As per other comments, I'm not sure this test is super useful. With the way the SDK is at the moment, it's not possible to achieve a result where you have a user or not if you're not authenticated, as that observable only emits something when you are authenticated (you can't have a user otherwise - conceptually it doesn't make sense). I think that makes this test hard to parse; at the beginning we're creating a user object then spying on `getUser` but none of that will be called.

What's the reason for adding this test in relation to `idTokenClaims`? Does it fill a gap that we didn't otherwise have?

Comment (frederikprijck) [reply projects/auth0-angular/src/lib/auth.service.spec.ts]: I think it comes from the discussion here https://github.com/auth0/auth0-angular/pull/69/files#discussion_r509113502 where we talked about verifying nothing gets emitted when not authenticated. 

Maybe this is something we should leave out of the current PR as I think it is worth discussing if we want to have the user$ and idTokenClaims$ emit null when not authenticated.

Doing so would result in the following observables:

**User Is Authenticated**
- user$ emits valid user
- idTokenClaims$ emits valid claims

**User logs out locally only**
- user$ emits null
- idTokenClaims$ emits null

Currently, when the user logsout locally, the user$ and idTokenClaims$ do not reflect that, which might be unexpected.

Comment (7opf) [reply projects/auth0-angular/src/lib/auth.service.spec.ts]: `user$` and `idTokenClaims$` both should emit when authenticated (positive) and not emit when not authenticated (negative). This test is testing the negative scenario, the other test above it tests the positive scenario. That was my thinking, but happy for you to remove or change as you see fit. And yes, spying on `getUser` is redundant - I was thinking that even if `getUser` were to emit, `user$` still shouldn't, but I see now this would never happen.

Comment (stevehobbsdev) [reply projects/auth0-angular/src/lib/auth.service.spec.ts]: In that case let's remove this test and potentially cover it elsewhere.

**Review by frederikprijck — APPROVED** [review]
